sw/device/mask_rom/mask_rom_start.S

```diff
@@ -0,0 +1,283 @@
+  .section .crt, "ax"
+
+  // Linker Relaxation is disabled until `__global_pointer$` is setup, below,
+  // because otherwise some sequences may be turned into gp-relative sequences,
+  // which is incorrect when `gp` is not initialized.
+  .option push
+  .option norelax
+
+/**
+ * Entry point after reset. This symbol is jumped to from the handler
+ * for IRQ 32.
+ *
+ * then jumps to `mask_rom_boot
+ */
+  .globl _mask_rom_start_boot
+  .type _mask_rom_start_boot, @function
+_mask_rom_start_boot:
+
+  /**
+   * Disable Interrupts.
+   *
+   * We cannot disable exceptions, or Ibex's non-maskable interrupts (interrupt
+   * 31), so we still need to be careful.
+   */
+  // Clear `MIE` field of `mstatus`.
+  csrci mstatus, 0x8
+  // Clear all the machine-defined interrupts, `MEIE`, `MTIE`, and `MSIE` fields
+  // of `mie`.
+  li   t0, 0xFFFF0888
+  csrc mie, t0
+
+  // Set up the stack pointer.
+  //
+  // If an exception fires, the handler is conventionaly only allowed to clobber
+  // memory at addresses below `sp`.
+  la   sp, _stack_start
+
+  /**
+   * Set well-defined interrupt/exception handlers
+   *
+   * We actually booted with this value of `mtvec`, so we shouldn't need to do
+   * this.
+   *
+   * The lowest two bits should be `0b01` to ensure we use vectored interrupts.
+   */
+  la   t0, _mask_rom_interrupt_vector
+  andi t0, t0, -4
+  ori  t0, t0, 0b01
+  csrw mtvec, t0
+
+  /**
+   * Clean Device State Part 1
+   */
+
+  // Zero all writable registers except x2 (sp).
+  mv x1,  zero
+  // NOT x2 (sp) - We have already set it to the right value above.
+  mv x3,  zero
+  mv x4,  zero
+  mv x5,  zero
+  mv x6,  zero
+  mv x7,  zero
+  mv x8,  zero
+  mv x9,  zero
+  mv x10, zero
+  mv x11, zero
+  mv x12, zero
+  mv x13, zero
+  mv x14, zero
+  mv x15, zero
+  mv x16, zero
+  mv x17, zero
+  mv x18, zero
+  mv x19, zero
+  mv x20, zero
+  mv x21, zero
+  mv x22, zero
+  mv x23, zero
+  mv x24, zero
+  mv x25, zero
+  mv x26, zero
+  mv x27, zero
+  mv x28, zero
+  mv x29, zero
+  mv x30, zero
+  mv x31, zero
+
+  // TODO: Setup SRAM Scrambling
+  // Temporarily: Zero out ram_main
+  //
+  // `t0` is the address to start zeroing at.
+  // `t1` is the address to stop zeroing at.
+  li  t0, TOP_EARLGREY_RAM_MAIN_BASE_ADDR
+  li  t1, (TOP_EARLGREY_RAM_MAIN_BASE_ADDR + TOP_EARLGREY_RAM_MAIN_SIZE_BYTES)
+  bge t0, t1, sram_zero_end
+sram_zero_loop:
+  // TODO: Unroll loop
+  sw   zero, 0x0(t0)
+  addi t0, t0, 0x4
+  blt  t0, t1, sram_zero_loop
+sram_zero_end:
+
+  /**
+   * Setup C Runtime
+   */
+
+  // Initialize the `.data` section.
+  //
+  // `t0` is the start pointer of `.data` (in RAM)
+  // `t1` is the end pointer of `.data` (in RAM)
+  // `t2` is the start pointer of `.data` (in ROM).
+  // `t3` is a scratch register for the copy.
+  la  t0, _data_start
+  la  t1, _data_end
+  la  t2, _data_init_start
+  bge t0, t1, data_copy_loop_end
+data_copy_loop:
+  // TODO: Unroll this loop
+  lw   t3, 0(t2)
+  sw   t3, 0(t0)
```

Comment:
Minor: Other loops use `0x0(t0)`, it would be nice to use either `0(t0)` or `0x0(t0)` in all loops.

Reply:
Yarp, now using `0x0`